core/build.gradle

```diff
@@ -62,36 +134,30 @@ dependencies {
 ext.dependenciesPrefix = 'com.microsoft.applicationinsights.core.dependencies'
 shadowJar {
     classifier=''
+    mergeServiceFiles()
```

Comment:
TLDR: there's a java feature to query for an implementation of an interface by creating a file in `META-INF/services` whose filename is the FQCN of the interface, inside is the FQCN of the implementation.

Here's the docs about [SPI](https://docs.oracle.com/javase/tutorial/ext/basics/spi.html)
Here's one of the classes which loads the implementations: https://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html

Anyway, since the classes referenced in these files are repackaged, we have to update the FQCNs of the filename and the files' contents.

For grpc to work, you must provide either the `grpc-netty` or the `grpc-okhttp` jar so it knows how to make connections.

Reply:
Thanks for the explanation. For the grpc, I would prefer grpc-netty over okhttp. Netty is non blocking and is performant.